# Chapter 0. 기본 중의 기본
## Pandas는 무엇인가??
#### 왜 pandas를 사용할까?
* pandas 사이트에 있는 소개  
  pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,built on top of the Python programming language.  
  파파고 : 팬더는 빠르고 강력하며 유연하며 사용하기 쉬운 Python 프로그래밍 언어 위에 구축된 오픈 소스 데이터 분석 및 조작 도구다.
  
    
* 특징(발번역..)
in-memory 데이터 구조와 CSV 및 텍스트 파일, Microsoft Excel, SQL 데이터베이스 및 고속 HDF5 형식 같은 데이터 형식 읽고 쓰기 가능  
지능형 데이터 정렬 및 누락 데이터의 통합 처리: 연산 시 자동 label-based 정렬, 데이터를 정렬형태로 쉽게 조작가능  
데이터 구조에서 열을 삽입 및 삭제 가능(가변성)  
고성능 데이터 세트 병합 및 결합 등등  
**요약하자면 큰 사이즈의 데이터를 조작하기 쉽다는 장점, 오픈소스 그리고 python으로 되어있다.**

---
### pandas 라이브러리 불러오기  
pandas는 pd로 축약해서 쓰는게 국룰

In [1]:
import pandas as pd

---
### 데이터프레임 불러오기

pd.read_csv([파일경로/파일이름]) 으로 불러오자  
**read_csv의 옵션들은 <code>pd.read_csv?</code>나 <code>pd.read_csv(에서 shift+tab+tab</code> 으로 확인가능**

In [2]:
# 실습 : df 변수에 파일을 불러와서 할당하기
# 불러올 파일 : ./data/5월_대전_아파트(매매)실거래가.csv  
# encoding="ANSI"  (인코딩)
# sep=","  (구분자)
df = pd.read_csv('./data/5월_대전_아파트(매매)실거래가.csv', encoding="ANSI",sep=',')

파일을 불러왔으면 봐야하는데 어떻게 봐야할까??
1. df (전체 표시, 그러나 길면 일부는 생략됨)  
2. df.head(n) (위에서부터 n개 표시, 기본값 5)
3. df.tail(n) (끝에서부터 n개 표시, 기본값 5)

In [3]:
# 실습 : df를 위에서부터 5개를 표시해보자
df.tail()

시군구     번지   본번  부번  단지명  전용면적(㎡)    계약년월  계약일 거래금액(만원)   층  \
2301  대전광역시 중구 태평동  389-5  389   5  파라곤   84.236  202005   16   36,000   6   
2302  대전광역시 중구 태평동  389-5  389   5  파라곤  106.375  202005   23   43,000  16   
2303  대전광역시 중구 태평동  254-4  254   4  푸른뫼   84.990  202005   16   29,500   8   
2304  대전광역시 중구 태평동  254-4  254   4  푸른뫼   84.990  202005   23   29,700  21   
2305  대전광역시 중구 태평동  254-4  254   4  푸른뫼   84.990  202005   28   27,000  16   

      건축년도        도로명  
2301  2005  유등천동로 428  
2302  2005  유등천동로 428  
2303  1999    수침로 122  
2304  1999    수침로 122  
2305  1999    수침로 122

---
### 데이터프레임 컬럼 확인하기  
[데이터프레임].columns 로 확인하자

In [4]:
# 실습 : df의 컬럼들의 목록을 확인해보자
df.columns

Index(['시군구', '번지', '본번', '부번', '단지명', '전용면적(㎡)', '계약년월', '계약일', '거래금액(만원)',
       '층', '건축년도', '도로명'],
      dtype='object')

# Chapter 1. 데이터프레임 깔짝대기

### 컬럼이름 변경하기  
df.rename을 사용해서 바꿔보자  
rename? 으로 옵션을 확인할 수 있음  
column이름을 바꿀거니까 rename(columns={})를 사용할 건데, 바꿀 컬럼이름들은 dict형태로 넘겨줌  
<code>df.rename(columns={'기존이름':'바꿀이름'})</code>

In [5]:
# 실습 : 컬럼 이름을 바꿔보자
# 바꿀 내용들 : "전용면적(㎡)" -> "전용면적"  
#               "거래금액(만원)" -> "거래금액"
df.rename(columns={'전용면적(㎡)':'전용면적','거래금액(만원)' : '거래금액' })

시군구     번지   본번  부번   단지명      전용면적    계약년월  계약일    거래금액   층  \
0     대전광역시 대덕구 대화동   27-7   27   7    금강   56.2600  202005   25   7,000   5   
1     대전광역시 대덕구 대화동   16-2   16   2  금성백조   63.0750  202005   11   8,700   3   
2     대전광역시 대덕구 대화동   16-2   16   2  금성백조   70.4250  202005   26   9,000   5   
3     대전광역시 대덕구 대화동  40-67   40  67    대화   49.0000  202005   12   3,400   3   
4     대전광역시 대덕구 대화동  31-36   31  36    두레   35.8818  202005    4   7,500   2   
...             ...    ...  ...  ..   ...       ...     ...  ...     ...  ..   
2301   대전광역시 중구 태평동  389-5  389   5   파라곤   84.2360  202005   16  36,000   6   
2302   대전광역시 중구 태평동  389-5  389   5   파라곤  106.3750  202005   23  43,000  16   
2303   대전광역시 중구 태평동  254-4  254   4   푸른뫼   84.9900  202005   16  29,500   8   
2304   대전광역시 중구 태평동  254-4  254   4   푸른뫼   84.9900  202005   23  29,700  21   
2305   대전광역시 중구 태평동  254-4  254   4   푸른뫼   84.9900  202005   28  27,000  16   

      건축년도        도로명  
0     1978   대전로 1247  
1     1985     대화로 19  
2     1985     대화로 19  
3     1979     대화로 53  
4     2001    대화1길 77  
...    ...        ...  
2301  2005  유등천동로 428  
2302  2005  유등천동로 428  
2303  1999    수침로 122  
2304  1999    수침로 122  
2305  1999    수침로 122  

[2306 rows x 12 columns]

In [6]:
# df.colums띄워보기
df.columns

Index(['시군구', '번지', '본번', '부번', '단지명', '전용면적(㎡)', '계약년월', '계약일', '거래금액(만원)',
       '층', '건축년도', '도로명'],
      dtype='object')

아까 바꾼거같죠 ㅋ 근데 사실 안바뀜 ㅋ 보여지기는 바꿔서 보여주지만 재할당을 하거나, 적용을 시켜주는 옵션을 넣어줘야함
1. <code>df.rename(columns={'전용면적(㎡)':'전용면적','거래금액(만원)' : '거래금액' }, inplace=True)</code>  
2. <code>df = df.rename(columns={'전용면적(㎡)':'전용면적','거래금액(만원)' : '거래금액' })</code>

In [7]:
# 실습 : 컬럼 이름 변경을 할당하게끔 다시 시도
df = df.rename(columns={'전용면적(㎡)':'전용면적','거래금액(만원)' : '거래금액' })
# 실습 : 컬럼목록 재확인하기
df.columns

Index(['시군구', '번지', '본번', '부번', '단지명', '전용면적', '계약년월', '계약일', '거래금액', '층',
       '건축년도', '도로명'],
      dtype='object')

---
### 인덱스 조작하기
보통 row조작, column조작, 둘다 조작이 있지만, row는 이번 프로젝트에서 쓴 경험이 없다..!  
그러므로 column 조작만 해봅시다..
* 특정 컬럼이름으로 불러오기  
  - 1개 컬럼  
   <code> df['컬럼이름'] </code>
  - 여러개 컬럼  
   <code> df[['컬럼1','컬럼2']]</code>  
   
   
* 조건을 만족하는 값 불러오기  
(데이터프레임[컬럼명] 조건) 이렇게 적으면 각 행에 대해 조건 만족여부(true,false)가 나오므로,  
데이터프레임으로 한번 더 싸서 **해당조건을 만족하는 값**들만 나타나게 해야함
  - 층이 10층 이상인 경우  
   <code> df[df['층'] >=10] </code>
  - 층이 10층 이상이고, 20층 이하인 경우  
  조건이 2개이상인 경우 연산자 우선순위때문에 ()로 묶어주기  
   <code> df[(df['층'] >=10) & (df['층']<=20)]</code>
  - 값이 없는 칼럼은??  
    <code>df[df['시군구'].isnull()]</code>
  - 값이 있는 칼럼은??  
    <code>df[df['시군구'].notnull()]</code>
    
str타입의 칼럼의 경우, str.contains(), str.split() 등을 쓸 수 있다

In [8]:
# 실습 : 건축년도가 2000 보다 큰 경우
df[df['건축년도']>2000]

시군구      번지   본번  부번      단지명      전용면적    계약년월  계약일    거래금액  \
4     대전광역시 대덕구 대화동   31-36   31  36       두레   35.8818  202005    4   7,500   
36    대전광역시 대덕구 목상동     204  204   0  두레마을 2차   59.9945  202005   21  14,200   
39    대전광역시 대덕구 목상동  207-25  207  25    평화로운    59.9788  202005    4  14,600   
40    대전광역시 대덕구 목상동  207-25  207  25    평화로운    59.9788  202005    4  14,000   
41    대전광역시 대덕구 목상동  207-25  207  25    평화로운    59.9788  202005   23  13,900   
...             ...     ...  ...  ..      ...       ...     ...  ...     ...   
2298   대전광역시 중구 태평동   389-5  389   5      파라곤   84.2360  202005    2  38,500   
2299   대전광역시 중구 태평동   389-5  389   5      파라곤   75.9340  202005    8  33,500   
2300   대전광역시 중구 태평동   389-5  389   5      파라곤   84.2360  202005   12  37,500   
2301   대전광역시 중구 태평동   389-5  389   5      파라곤   84.2360  202005   16  36,000   
2302   대전광역시 중구 태평동   389-5  389   5      파라곤  106.3750  202005   23  43,000   

       층  건축년도            도로명  
4      2  2001        대화1길 77  
36     1  2014  대덕대로1486번길 65  
39     6  2011  대덕대로1520번길 77  
40    12  2011  대덕대로1520번길 77  
41     2  2011  대덕대로1520번길 77  
...   ..   ...            ...  
2298  17  2005      유등천동로 428  
2299   7  2005      유등천동로 428  
2300  20  2005      유등천동로 428  
2301   6  2005      유등천동로 428  
2302  16  2005      유등천동로 428  

[903 rows x 12 columns]

In [9]:
# 실습 : 시군구에 대덕구가 포함된 경우
df[df['시군구'].str.contains('대덕구')]

시군구     번지   본번  부번    단지명     전용면적    계약년월  계약일    거래금액   층  \
0    대전광역시 대덕구 대화동   27-7   27   7     금강  56.2600  202005   25   7,000   5   
1    대전광역시 대덕구 대화동   16-2   16   2   금성백조  63.0750  202005   11   8,700   3   
2    대전광역시 대덕구 대화동   16-2   16   2   금성백조  70.4250  202005   26   9,000   5   
3    대전광역시 대덕구 대화동  40-67   40  67     대화  49.0000  202005   12   3,400   3   
4    대전광역시 대덕구 대화동  31-36   31  36     두레  35.8818  202005    4   7,500   2   
..             ...    ...  ...  ..    ...      ...     ...  ...     ...  ..   
387  대전광역시 대덕구 평촌동  540-1  540   1  행복한마을  66.2681  202005    4  19,500   7   
388  대전광역시 대덕구 평촌동  540-1  540   1  행복한마을  66.2681  202005    8  19,800  12   
389  대전광역시 대덕구 평촌동  540-1  540   1  행복한마을  84.7051  202005    9  22,850  14   
390  대전광역시 대덕구 평촌동  540-1  540   1  행복한마을  84.5087  202005   17  21,700   8   
391  대전광역시 대덕구 평촌동  540-1  540   1  행복한마을  84.7154  202005   17  23,200  16   

     건축년도       도로명  
0    1978  대전로 1247  
1    1985    대화로 19  
2    1985    대화로 19  
3    1979    대화로 53  
4    2001   대화1길 77  
..    ...       ...  
387  2011   덕암로 200  
388  2011   덕암로 200  
389  2011   덕암로 200  
390  2011   덕암로 200  
391  2011   덕암로 200  

[392 rows x 12 columns]

In [10]:
# 실습 : 시군구가 없는 경우
df[df['시군구'].isnull()]

Empty DataFrame
Columns: [시군구, 번지, 본번, 부번, 단지명, 전용면적, 계약년월, 계약일, 거래금액, 층, 건축년도, 도로명]
Index: []

### 데이터프레임 컬럼 타입 확인하기
[데이터프레임저장한 변수].dtypes 으로 확인하자  
dtype에서 object가 str이라고 생각하면 된다

In [11]:
df.dtypes

시군구      object
번지       object
본번        int64
부번        int64
단지명      object
전용면적    float64
계약년월      int64
계약일       int64
거래금액     object
층         int64
건축년도      int64
도로명      object
dtype: object

## 계약년월+계약일 으로 계약날짜 칼럼 만들기
---

In [12]:
# 실행에 앞서 첫 줄만 띄워보자
# 실습 : df의 첫번째 row만 출력하기 (힌트 : head() 이용)
df.head(1)

시군구    번지  본번  부번 단지명   전용면적    계약년월  계약일   거래금액  층  건축년도  \
0  대전광역시 대덕구 대화동  27-7  27   7  금강  56.26  202005   25  7,000  5  1978   

        도로명  
0  대전로 1247

### 컬럼 값 확인하기
1. 중복없이 값 확인하기 - unique()  
<code>df['컬럼명'].unique</code>
2. 값이 몇개씩 있는지 확인하기 - value_counts()  
<code> df['컬럼명'].value_counts()</code>

In [13]:
# 실습 : 시군구가 몇개씩 있는지 확인하기
df['시군구'].value_counts()

대전광역시 서구 관저동     151
대전광역시 대덕구 법동     127
대전광역시 유성구 전민동    118
대전광역시 유성구 지족동     95
대전광역시 서구 둔산동      88
                ... 
대전광역시 동구 신흥동       2
대전광역시 동구 중동        2
대전광역시 서구 괴정동       1
대전광역시 대덕구 신일동      1
대전광역시 중구 문창동       1
Name: 시군구, Length: 89, dtype: int64

In [14]:
# 실습 : 계약일이 중복없이 무슨 값이 있는지 확인하기
df['계약일'].unique()

array([25, 11, 26, 12,  4,  3,  6, 15, 18, 19, 20,  8, 22,  9, 21, 23,  1,
       14, 30,  2, 13, 27,  7, 16, 28, 24, 29, 10,  5, 31, 17],
      dtype=int64)

#### 한자리 수를 가진 일(날짜)에 0을 추가해야겠다..

### 데이터타입 변경하기  
[계약년월]과 [계약일]이 int타입이므로 다루기 쉽게 str(object)로 바꾸자  
astype을 이용해서 rename-columns처럼 dict형식으로 주면됨   
<code>astype({'컬럼이름':타입, '컬럼이름1':타입})</code>

In [15]:
# 실습 : 계약년월, 계약일 컬럼을 str 타입으로 바꾸자  
# 질문, 이거는 바로 반영될까, 아니면 아까처럼 할당을 따로 해줘야할까?
df = df.astype({'계약년월':str, '계약일':str})

In [16]:
# 확인하기
df.dtypes

시군구      object
번지       object
본번        int64
부번        int64
단지명      object
전용면적    float64
계약년월     object
계약일      object
거래금액     object
층         int64
건축년도      int64
도로명      object
dtype: object

### 계약일이 2자리 수로 맞춰있는 경우라면?? 훨신 간단했을텐데..  
<code>df['계약날짜'] = df['계약년월']+df['계약일']</code>

In [17]:
df['계약날짜'] = df['계약년월']+df['계약일']

In [18]:
df['계약날짜'].unique()

array(['20200525', '20200511', '20200526', '20200512', '2020054',
       '2020053', '2020056', '20200515', '20200518', '20200519',
       '20200520', '2020058', '20200522', '2020059', '20200521',
       '20200523', '2020051', '20200514', '20200530', '2020052',
       '20200513', '20200527', '2020057', '20200516', '20200528',
       '20200524', '20200529', '20200510', '2020055', '20200531',
       '20200517'], dtype=object)

하지만 **2020053** 같은 결과를 얻고싶지않아..  
한자리 수 날짜인 경우에는 0을 추가하는 함수를 만들자!

### apply 사용하기  
def 함수(row):
    함수 로직
    
할당까지 마쳐줘야함 ! 근데 다중칼럼 적용은 안되는 걸로 알고있음..
* 함수를 특정 칼럼값만 넘겨주고 싶을 때  
  - <code>df[새 칼럼] = df[컬럼명].apply(함수이름)</code>
  
* 함수를 row전체를 넘겨주고 싶을 때  
  - <code>df[새 칼럼] = df.apply(함수이름,axis=1)</code>

In [19]:
# 함수 결과 확인용
df['계약일'].unique()

array(['25', '11', '26', '12', '4', '3', '6', '15', '18', '19', '20', '8',
       '22', '9', '21', '23', '1', '14', '30', '2', '13', '27', '7', '16',
       '28', '24', '29', '10', '5', '31', '17'], dtype=object)

In [20]:
# 실습 : 계약일 칼럼에 def addZero()함수를 통과시켜서 한자리 숫자인경우에는 0을 붙여보자 !!
# 1. 함수 정의하기  
# 내용을 채워보자
def addZero(day):
    if len(day)==1:
        return "0"+day
        pass
    else:
        return day

In [21]:
# 실습 : df['계약일'] 칼럼에 addZero를 apply하는 내용을 적어보자
df['계약일'] = df['계약일'].apply(addZero)

In [22]:
# 결과 :
df['계약일'].unique()

array(['25', '11', '26', '12', '04', '03', '06', '15', '18', '19', '20',
       '08', '22', '09', '21', '23', '01', '14', '30', '02', '13', '27',
       '07', '16', '28', '24', '29', '10', '05', '31', '17'], dtype=object)

In [23]:
# 실습 : 계약날짜를 계약년월 + 바꾼 계약일 로 바꾸기
df['계약날짜'] = df['계약년월']+df['계약일']

**만약 apply함수에 row를 줬다면, row['계약일'] 이런식으로 특정 칼럼의 값을 전달할 수 있음**

In [24]:
df.head(1)

시군구    번지  본번  부번 단지명   전용면적    계약년월 계약일   거래금액  층  건축년도  \
0  대전광역시 대덕구 대화동  27-7  27   7  금강  56.26  202005  25  7,000  5  1978   

        도로명      계약날짜  
0  대전로 1247  20200525

## 시군구 + 도로명으로 도로명주소 칼럼을 만들어보기(실습)
---

In [25]:
# 실습 : 혹시 도로명주소가 없는게 있을까? isnull()로 확인하기
df[df['도로명'].isnull()]

Empty DataFrame
Columns: [시군구, 번지, 본번, 부번, 단지명, 전용면적, 계약년월, 계약일, 거래금액, 층, 건축년도, 도로명, 계약날짜]
Index: []

In [26]:
# 실습 : ['시군구']에서 동을 제외한(시,구) 내용과 ['도로명']을 합쳐서 새 칼럼인 ['도로명주소']를 만들어보자
# 조건 : apply 사용(row 넘겨주기)
# split(" ")
def getNewAddress(row):
    t = row['시군구'].split(" ")
    return t[0]+" "+t[1]+" "+row['도로명']
    
df['도로명주소'] = df.apply(getNewAddress,axis=1)

### 위도경도 받아오기(참고)

### 카카오 API 써서 주소가 덜 작성된거는 위도 경도로 새 주소로 업뎃해쥼
```
import requests
import json

headers = {
"Authorization":"KakaoAK 66447c8b7057c042f65ac177c53ea385"
}

address = {}
def getLatLong(addr):
    if addr in address:
        return address[addr]
    else:
        params = {'query': addr}
        response = requests.get('https://dapi.kakao.com/v2/local/search/address.json', headers=headers, params=params)
        count = response.json()['meta']['total_count']
        if count!=0:
            res = response.json()
            try:
                lat =res['documents'][0]['address']['y']
                lon = res['documents'][0]['address']['x']
                addr1 = res['documents'][0]['address']['region_1depth_name']
                addr2 = res['documents'][0]['address']['region_2depth_name']
                addr3 = res['documents'][0]['address']['region_3depth_name']
                address[addr] = str(lon)+"/"+str(lat)
                return address[addr]
            except:
                print(res)
                return "ERROR"
        else :
            return "NO"

df['latlon'] = df['도로명주소'].apply(getLatLong)
```

---
## 파일 저장하기
저장할데이터프레임.to_csv(경로+파일명,옵션) 으로 저장할 수 있다  
옵션같은 경우는 마찬가지로 to_csv?나, to_csv(+shift+tab 으로 확인

In [27]:
# 실습 : df를 새 파일로 인덱스 없이 저장하자
# 힌트 : 인덱스 없이??? 옵션한번 볼까??
df.to_csv('./data/new_df.csv', index=False)

# Chapter 2. mysql활용하기
사용 라이브러리 :  
```python
import pymysql
from sqlalchemy import create_engine
```
만일 없다면,, :
```python
! pip install PyMySQL
! pip install SQLAlchemy
```

In [28]:
# 에러가 없어야함
import pymysql
from sqlalchemy import create_engine

In [29]:
# 설치 필요한 경우


### DB 연결
```python
db_connection_str = 'mysql+pymysql://[계정이름]:[비밀번호]@[주소]:[포트]/[데이터베이스이름]'
db_connection = create_engine(db_connection_str)
conn = db_connection.connect()
```
내가 데이터베이스 hongju에 daejeon 테이블 만들어놨다 !!!  
```
CREATE TABLE `hongju`.`daejeon` (
  `시군구` VARCHAR(45) NULL,
  `num` INT NOT NULL AUTO_INCREMENT,
  `전용면적` VARCHAR(45) NULL,
  `계약날짜` DATE NULL,
  `도로명주소` VARCHAR(300) NULL,
  `층` VARCHAR(45) NULL,
  `건축년도` INT NULL,
  PRIMARY KEY (`num`));
```

In [30]:
# 실습 : 우리 서버디비에 hongju데이터베이스 접속하기
db_connection_str = 'mysql+pymysql://test:rot123456789!!@k02b2031.p.ssafy.io:3306/hongju'
db_connection = create_engine(db_connection_str)
conn = db_connection.connect()

현재 daejeon테이블에 보면 
**num, 시군구, 전용면적, 계약날짜, 도로명주소, 층, 건축년도** 가 있어요

지금 우리가 갖고있는 df는
<code>df.dtypes</code>를 하면 아래와같이 나옵니다.
```
시군구        object
번지         object
본번          int64
부번          int64
단지명        object
전용면적      float64
계약년월       object
계약일        object
거래금액       object
층           int64
건축년도        int64
도로명        object
계약날짜       object
도로명주소      object
dtype: object
```

In [31]:
# 실습 : selected_df에 시군구, 전용면적, 계약날짜, 도로명주소, 층, 건축년도  칼럼만 담아보자 !
selected_df = df[['시군구','전용면적','계약날짜','도로명주소','층','건축년도']]

In [32]:
selected_df

시군구      전용면적      계약날짜               도로명주소   층  건축년도
0     대전광역시 대덕구 대화동   56.2600  20200525  대전광역시 대덕구 대전로 1247   5  1978
1     대전광역시 대덕구 대화동   63.0750  20200511    대전광역시 대덕구 대화로 19   3  1985
2     대전광역시 대덕구 대화동   70.4250  20200526    대전광역시 대덕구 대화로 19   5  1985
3     대전광역시 대덕구 대화동   49.0000  20200512    대전광역시 대덕구 대화로 53   3  1979
4     대전광역시 대덕구 대화동   35.8818  20200504   대전광역시 대덕구 대화1길 77   2  2001
...             ...       ...       ...                 ...  ..   ...
2301   대전광역시 중구 태평동   84.2360  20200516  대전광역시 중구 유등천동로 428   6  2005
2302   대전광역시 중구 태평동  106.3750  20200523  대전광역시 중구 유등천동로 428  16  2005
2303   대전광역시 중구 태평동   84.9900  20200516    대전광역시 중구 수침로 122   8  1999
2304   대전광역시 중구 태평동   84.9900  20200523    대전광역시 중구 수침로 122  21  1999
2305   대전광역시 중구 태평동   84.9900  20200528    대전광역시 중구 수침로 122  16  1999

[2306 rows x 6 columns]

### how to 데이터베이스에 데이터 프레임넣기
```python
df.to_sql(name='db의 테이블이름', con=db_connection, if_exists='append',index=False)  
```

**con** : con에 연결되는 db_connection은 create_engine()를 담은 변수를 써주면 된다. 


**if_exists** : [fail, replace, append] -> 테이블이 이미 존재하는 경우 어떻게 할지에 대한 옵션  


**fail** : ValueError 발생  


**append** : 존재하는 테이블에 값 저장  
append의 경우는 기존에 있는 테이블에 저장을 하는 것이여서 저장할 데이터프레임의 길이나 타입이 맞지않는다면 오류가 난다.  


**replace** : 테이블 지우고 새로 생성 후 값 저장  
replace는 새로 데이터프레임 타입에 맞춰 테이블이 생성되므로 데이터프레임 타입이 object일땐 mysql에서는 TEXT로 생겼다.  
dtype옵션을 줘서 타입을 맞춰줄 수 있다.  

In [33]:
# 실습, daejeon테이블에 selected_df를 저장해보기
selected_df.to_sql(name='daejeon', con=db_connection, if_exists='append',index=False)

---
### 데이터베이스에서 테이블을 데이터프레임으로 불러오기
db연결을 아직 안끊었으니까 around테이블을 불러오자 !
```python
pd.read_sql("SELECT * from [특정테이블명]", con=conn)
```
이제 쓸 일 없으니까 <code>conn.close()</code>로 꼭꼭 연결끊기 !

In [34]:
# 실습 : around 테이블을 데이터베이스에서 불러와서 데이터프레임으로 변환한 후, around 변수에 할당하기
around = pd.read_sql("SELECT * from around", con=conn)

In [35]:
# 실습 : db연결 종료하기
conn.close()

# Chapter 3. 심화1 - 그룹화
지금 우리 수중에 있는 데이터프레임은 2개  
1. db에 막 업로드한 selected_df  
2. db에서 불러온 around  

심화에서 할 내용 :  
1. selected_df와 around테이블을 도로명주소, address로 묶어보자
2. 시군구로 묶어 각 시군구에 몇개의 around가 속해있는지 보자

### step 1. selected_df와 around테이블을 도로명주소, address로 묶어보자
>Q. 선생님 어떻게 묶나요.. 무슨소린지도 모르겠어요..  
>A. pd.merge()를 이용하면 됩니다 !  
>Q. ???!  

데이터프레임 두개를 묶을 때는 2가지 방법이 있읍니다..  
1. concat : 동일한 index,혹은 column을 가진 경우 연속적으로 붙이기
2. merge : 공통된 하나의 열이나 행을 기준으로 동일한 값을 가지는 행을 dataframe에서 찾은 뒤, n x m개 조합으로 행을 모두 가지도록 합치는 경우  


나는 주로 merge를 쓰지 ~!  
merge는 집합을 생각하면 좋습니다 ! 
```python
pd.merge(zoo,zoo_eats, how='inner')
```
이런 방식으로 어느 데이터프레임들을 묶을지, 어떤방식을 사용할 지 ! 적으면 됩니다.
left_on, right_on을 적어주지않으면 기본적으로 같은 이름을 가지는 컬럼들을 기준으로 진행합니다.(아마도)

![img](https://data36.com/wp-content/uploads/2018/08/4-pandas-merge-inner-outer-left-right-1024x771.png)

---
concat : https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html  
merge : https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html  

---



select_df - 도로명주소  
around - address 칼럼을 기준으로 merge를 해봅시다 !

In [36]:
# 실습 : merge해서 t 변수에 담기
t = pd.merge(selected_df, around, left_on='도로명주소',right_on="address", how='inner')

In [37]:
# 실습 : 가장 위의 1개 row만 가져오기
t.head(1)

시군구    전용면적      계약날짜             도로명주소  층  건축년도  around_id  \
0  대전광역시 대덕구 대화동  63.075  20200511  대전광역시 대덕구 대화로 19  3  1985      23814   

            address      longitude       latitude  trans  comforts  education  \
0  대전광역시 대덕구 대화로 19  127.417902548  36.3674348466      0        16         87   

   medical  eatery  culture  
0       30     278        5

### groupby해보기  

이제 t 데이터프레임에서 같은 시군구를 가진 around들이 몇개가 있는지 봐볼까요??  
기억이 날진 모르겠지만,, t에서 시군구를 가지고 갯수 카운트를 해봅시다

In [38]:
# 실습 : t 데이터프레임의 시군구를 가지고 몇개씩 있는지 확인해보기
t['시군구'].value_counts()

대전광역시 서구 관저동      63
대전광역시 유성구 송강동     58
대전광역시 대덕구 송촌동     45
대전광역시 유성구 관평동     42
대전광역시 중구 태평동      39
                  ..
대전광역시 대덕구 신탄진동     2
대전광역시 중구 대사동       1
대전광역시 중구 문창동       1
대전광역시 동구 대성동       1
대전광역시 동구 중동        1
Name: 시군구, Length: 70, dtype: int64

groupby는 말그대로 기준을 가지고 그룹을 짓는 거에요  
**데이터프레임.groupby('기준컬럼')** 이런식으로 합니다  
t를 시군구로 묶어볼게요

In [39]:
groupby_sgg = t.groupby('시군구')

In [40]:
groupby_sgg.groups

{'대전광역시 대덕구 대화동': Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], dtype='int64'),
 '대전광역시 대덕구 덕암동': Int64Index([16, 17, 18, 19], dtype='int64'),
 '대전광역시 대덕구 목상동': Int64Index([20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30], dtype='int64'),
 '대전광역시 대덕구 법동': Int64Index([31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45], dtype='int64'),
 '대전광역시 대덕구 비래동': Int64Index([46, 47, 48, 49], dtype='int64'),
 '대전광역시 대덕구 석봉동': Int64Index([50, 51, 52, 53], dtype='int64'),
 '대전광역시 대덕구 송촌동': Int64Index([54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70,
             71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87,
             88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98],
            dtype='int64'),
 '대전광역시 대덕구 신탄진동': Int64Index([99, 100], dtype='int64'),
 '대전광역시 대덕구 연축동': Int64Index([101, 102, 103, 104, 105, 106, 107, 108], dtype='int64'),
 '대전광역시 대덕구 와동': Int64Index([109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121,
          

>Q. 엥.. 이거 보기에 너무 안좋은데요 ... 제가필요한건 그냥 그룹의 개수에요 ...  
>A. 그럴 땐 agg를 쓰면 됩니다 !  
>Q. ???!  


함수는 mean, size, min, max 등 기본제공되는 함수가있고, 내가만든 함수를 대입할수도 있어요
```python
groupby_sgg.agg(count=('대상칼럼','함수이름')) # count라는 이름의 컬럼에 넣어줄게
```



In [41]:
# 실습 : groupby_sgg를 around_id의 갯수로 그룹해서 count라는 이름의 컬럼에 출력
# 실습 : count 정렬하기
# 힌트 : 대상.sort_values(by="기준", ascending=False)
groupby_sgg.agg(count=('around_id','size')).sort_values(by="count", ascending=False)

count
시군구                  
대전광역시 서구 관저동       63
대전광역시 유성구 송강동      58
대전광역시 대덕구 송촌동      45
대전광역시 유성구 관평동      42
대전광역시 중구 태평동       39
...               ...
대전광역시 대덕구 신탄진동      2
대전광역시 중구 문창동        1
대전광역시 중구 대사동        1
대전광역시 동구 중동         1
대전광역시 동구 대성동        1

[70 rows x 1 columns]

# Chapter 4. 심화2 - 간단한 시각화(참고용)  
그래프를 그리기 위해 matplotlib을 추가합니다  
set_config는 한글폰트를 위한것이므로 가-무(가볍게 무시)  
seaborn쓰면 더 많이 그릴수있을텐데 생략.. 알아보세요 ;;;  
https://seaborn.pydata.org/tutorial.html

In [42]:
import matplotlib as plt
import matplotlib.font_manager as fm
def set_config():
    # 폰트, 그래프 색상 설정
    font_list = fm.findSystemFonts(fontpaths=None, fontext="ttf")
    if any(["notosanscjk" in font.lower() for font in font_list]):
        plt.rcParams["font.family"] = "Noto Sans CJK JP"
    else:
        if not any(["malgun" in font.lower() for font in font_list]):
            raise Exception(
                "Font missing, please install Noto Sans CJK or Malgun Gothic. If you're using ubuntu, try `sudo apt install fonts-noto-cjk`"
            )

        plt.rcParams["font.family"] = "Malgun Gothic"

    sns.set_palette(sns.color_palette("Spectral"))
    plt.rc("xtick", labelsize=6)
set_config()

NameError: name 'sns' is not defined

In [ ]:
plt.rcParams["figure.figsize"] = (10,15) # 차트 사이즈
t['시군구'].value_counts().plot(kind='barh')

In [ ]:
df

In [45]:
t

시군구     전용면적      계약날짜               도로명주소   층  건축년도  \
0    대전광역시 대덕구 대화동  63.0750  20200511    대전광역시 대덕구 대화로 19   3  1985   
1    대전광역시 대덕구 대화동  70.4250  20200526    대전광역시 대덕구 대화로 19   5  1985   
2    대전광역시 대덕구 대화동  35.8818  20200504   대전광역시 대덕구 대화1길 77   2  2001   
3    대전광역시 대덕구 대화동  71.8200  20200503  대전광역시 대덕구 대화11길 11   5  1987   
4    대전광역시 대덕구 대화동  54.5650  20200506  대전광역시 대덕구 대화11길 11   3  1987   
..             ...      ...       ...                 ...  ..   ...   
900   대전광역시 중구 태평동  57.3208  20200508     대전광역시 중구 수침로 42   3  1983   
901   대전광역시 중구 태평동  57.3208  20200511     대전광역시 중구 수침로 42   4  1983   
902   대전광역시 중구 태평동  84.9900  20200516    대전광역시 중구 수침로 122   8  1999   
903   대전광역시 중구 태평동  84.9900  20200523    대전광역시 중구 수침로 122  21  1999   
904   대전광역시 중구 태평동  84.9900  20200528    대전광역시 중구 수침로 122  16  1999   

     around_id             address      longitude        latitude  trans  \
0        23814    대전광역시 대덕구 대화로 19  127.417902548   36.3674348466      0   
1        23814    대전광역시 대덕구 대화로 19  127.417902548   36.3674348466      0   
2        23816   대전광역시 대덕구 대화1길 77  127.416590543  36.36391287420      0   
3        23817  대전광역시 대덕구 대화11길 11  127.417475544  36.36653552860      0   
4        23817  대전광역시 대덕구 대화11길 11  127.417475544  36.36653552860      0   
..         ...                 ...            ...             ...    ...   
900      24587     대전광역시 중구 수침로 42  127.401604586   36.3236824631      2   
901      24587     대전광역시 중구 수침로 42  127.401604586   36.3236824631      2   
902      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   
903      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   
904      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   

     comforts  education  medical  eatery  culture  
0          16         87       30     278        5  
1          16         87       30     278        5  
2          23         54       31     341        8  
3          15         76       25     288        7  
4          15         76       25     288        7  
..        ...        ...      ...     ...      ...  
900        56        263      152    1046       16  
901        56        263      152    1046       16  
902        36        207       68     542       11  
903        36        207       68     542       11  
904        36        207       68     542       11  

[905 rows x 16 columns]

In [43]:
t.drop()

시군구     전용면적      계약날짜               도로명주소   층  건축년도  \
0    대전광역시 대덕구 대화동  63.0750  20200511    대전광역시 대덕구 대화로 19   3  1985   
1    대전광역시 대덕구 대화동  70.4250  20200526    대전광역시 대덕구 대화로 19   5  1985   
2    대전광역시 대덕구 대화동  35.8818  20200504   대전광역시 대덕구 대화1길 77   2  2001   
3    대전광역시 대덕구 대화동  71.8200  20200503  대전광역시 대덕구 대화11길 11   5  1987   
4    대전광역시 대덕구 대화동  54.5650  20200506  대전광역시 대덕구 대화11길 11   3  1987   
..             ...      ...       ...                 ...  ..   ...   
900   대전광역시 중구 태평동  57.3208  20200508     대전광역시 중구 수침로 42   3  1983   
901   대전광역시 중구 태평동  57.3208  20200511     대전광역시 중구 수침로 42   4  1983   
902   대전광역시 중구 태평동  84.9900  20200516    대전광역시 중구 수침로 122   8  1999   
903   대전광역시 중구 태평동  84.9900  20200523    대전광역시 중구 수침로 122  21  1999   
904   대전광역시 중구 태평동  84.9900  20200528    대전광역시 중구 수침로 122  16  1999   

     around_id             address      longitude        latitude  trans  \
0        23814    대전광역시 대덕구 대화로 19  127.417902548   36.3674348466      0   
1        23814    대전광역시 대덕구 대화로 19  127.417902548   36.3674348466      0   
2        23816   대전광역시 대덕구 대화1길 77  127.416590543  36.36391287420      0   
3        23817  대전광역시 대덕구 대화11길 11  127.417475544  36.36653552860      0   
4        23817  대전광역시 대덕구 대화11길 11  127.417475544  36.36653552860      0   
..         ...                 ...            ...             ...    ...   
900      24587     대전광역시 중구 수침로 42  127.401604586   36.3236824631      2   
901      24587     대전광역시 중구 수침로 42  127.401604586   36.3236824631      2   
902      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   
903      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   
904      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   

     comforts  education  medical  eatery  culture  
0          16         87       30     278        5  
1          16         87       30     278        5  
2          23         54       31     341        8  
3          15         76       25     288        7  
4          15         76       25     288        7  
..        ...        ...      ...     ...      ...  
900        56        263      152    1046       16  
901        56        263      152    1046       16  
902        36        207       68     542       11  
903        36        207       68     542       11  
904        36        207       68     542       11  

[905 rows x 16 columns]

In [ ]:
t.drop(columns='도로명주소')

In [46]:
t.describe()

전용면적           층         건축년도     around_id       trans  \
count  905.000000  905.000000   905.000000    905.000000  905.000000   
mean    74.825547    8.751381  1999.190055  24193.417680    0.448619   
std     23.975867    5.525190     9.293715    279.868945    0.782737   
min     17.550000    1.000000  1978.000000  22220.000000    0.000000   
25%     59.760000    4.000000  1993.000000  24030.000000    0.000000   
50%     74.990000    8.000000  1998.000000  24191.000000    0.000000   
75%     84.930000   13.000000  2006.000000  24378.000000    1.000000   
max    181.786400   26.000000  2018.000000  25038.000000    3.000000   

         comforts   education     medical       eatery     culture  
count  905.000000  905.000000  905.000000   905.000000  905.000000  
mean    31.803315  146.712707   62.671823   527.921547    6.994475  
std     22.042773   89.445584   57.010532   355.161738    6.448140  
min      3.000000    6.000000    0.000000    50.000000    0.000000  
25%     18.000000   98.000000   30.000000   308.000000    4.000000  
50%     26.000000  123.000000   51.000000   456.000000    5.000000  
75%     37.000000  192.000000   78.000000   651.000000    9.000000  
max    122.000000  608.000000  390.000000  1953.000000   46.000000

In [47]:
t

시군구     전용면적      계약날짜               도로명주소   층  건축년도  \
0    대전광역시 대덕구 대화동  63.0750  20200511    대전광역시 대덕구 대화로 19   3  1985   
1    대전광역시 대덕구 대화동  70.4250  20200526    대전광역시 대덕구 대화로 19   5  1985   
2    대전광역시 대덕구 대화동  35.8818  20200504   대전광역시 대덕구 대화1길 77   2  2001   
3    대전광역시 대덕구 대화동  71.8200  20200503  대전광역시 대덕구 대화11길 11   5  1987   
4    대전광역시 대덕구 대화동  54.5650  20200506  대전광역시 대덕구 대화11길 11   3  1987   
..             ...      ...       ...                 ...  ..   ...   
900   대전광역시 중구 태평동  57.3208  20200508     대전광역시 중구 수침로 42   3  1983   
901   대전광역시 중구 태평동  57.3208  20200511     대전광역시 중구 수침로 42   4  1983   
902   대전광역시 중구 태평동  84.9900  20200516    대전광역시 중구 수침로 122   8  1999   
903   대전광역시 중구 태평동  84.9900  20200523    대전광역시 중구 수침로 122  21  1999   
904   대전광역시 중구 태평동  84.9900  20200528    대전광역시 중구 수침로 122  16  1999   

     around_id             address      longitude        latitude  trans  \
0        23814    대전광역시 대덕구 대화로 19  127.417902548   36.3674348466      0   
1        23814    대전광역시 대덕구 대화로 19  127.417902548   36.3674348466      0   
2        23816   대전광역시 대덕구 대화1길 77  127.416590543  36.36391287420      0   
3        23817  대전광역시 대덕구 대화11길 11  127.417475544  36.36653552860      0   
4        23817  대전광역시 대덕구 대화11길 11  127.417475544  36.36653552860      0   
..         ...                 ...            ...             ...    ...   
900      24587     대전광역시 중구 수침로 42  127.401604586   36.3236824631      2   
901      24587     대전광역시 중구 수침로 42  127.401604586   36.3236824631      2   
902      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   
903      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   
904      24592    대전광역시 중구 수침로 122  127.399453941   36.3303462199      1   

     comforts  education  medical  eatery  culture  
0          16         87       30     278        5  
1          16         87       30     278        5  
2          23         54       31     341        8  
3          15         76       25     288        7  
4          15         76       25     288        7  
..        ...        ...      ...     ...      ...  
900        56        263      152    1046       16  
901        56        263      152    1046       16  
902        36        207       68     542       11  
903        36        207       68     542       11  
904        36        207       68     542       11  

[905 rows x 16 columns]